ykllvmwrap/src/ykllvmwrap.cc

```diff
@@ -37,15 +56,113 @@ extern "C" char *__yk_llvmwrap_symbolizer_find_code_sym(
+    ExecutionEngine *EE = EngineBuilder(std::move(MPtr))
+        .setMemoryManager(std::unique_ptr<MCJITMemoryManager>(memman))
+        .create();
+	EE->finalizeObject();
+
+    if (EE->hasError())
+        errx(EXIT_FAILURE, "Couldn't compile trace: %s", EE->getErrorMessage().c_str());
+
+    return (void *) EE->getFunctionAddress(TRACE_FUNC_NAME);
+}
+
+// Compiles an IRTrace to executable code in memory.
 //
 // The trace to compile is passed in as two arrays of length Len. Then each
 // (FuncName[I], BBs[I]) pair identifies the LLVM block at position `I` in the
 // trace.
-extern "C" void __ykllvmwrap_irtrace_compile(char *FuncNames[], size_t BBs[], size_t Len) {
+//
+// Returns a pointer to the compiled function.
+extern "C" void *__ykllvmwrap_irtrace_compile(char *FuncNames[], size_t BBs[], size_t Len, char *SecPtr, size_t SecSize) {
+    LLVMContext Context;
+    auto DstMod = new Module("", Context);
+
+    // Set up new module for the trace.
+    llvm::FunctionType *FType = llvm::FunctionType::get(Type::getVoidTy(Context), false);
+	llvm::Function *DstFunc = llvm::Function::Create(FType, Function::InternalLinkage, TRACE_FUNC_NAME, DstMod);
+    DstFunc->setCallingConv(CallingConv::C);
+    auto DstBB = BasicBlock::Create(Context, "bb0", DstFunc);
+	llvm::IRBuilder<> Builder(Context);
+	Builder.SetInsertPoint(DstBB);
+
+    // Load the binary's bitcode.
+    auto SrcMod = load_module(Context, SecPtr, SecSize);
+
+    llvm::ValueToValueMapTy VMap;
+    bool Tracing = false;
+
+    // Iterate over the PT trace and stitch together all traced blocks.
     for (size_t Idx = 0; Idx < Len; Idx++) {
-        // FIXME populate.
-        //auto FuncName = FuncNames[Idx];
-        //auto BB = BBs[Idx];
+        auto FuncName = FuncNames[Idx];
+
+        // Get a traced function so we can extract blocks from it.
+        Function *F = SrcMod->getFunction(FuncName);
+        if (!F)
+            errx(EXIT_FAILURE, "can't find function %s", FuncName);
+
+        // Skip to the correct block.
+        auto It = F->begin();
+        std::advance(It, BBs[Idx]);
+        BasicBlock *BB = &*It;
+        // Iterate over all instructions within this block and copy them over
+        // to our new module.
+        for (auto I = BB->begin(), EI = BB->end(); I != EI; I++) {
+            if (isa<CallInst>(I)) {
+                Function *CF = cast<CallInst>(&*I)->getCalledFunction();
+                if (CF->getName() == YKTRACE_START) {
+                    Tracing = true;
+                    I++; // FIXME Better way of removing result of __yktrace_start_tracing.
+                    continue;
+                } else if (CF->getName() == YKTRACE_STOP) {
+                    // FIXME Remove argument setup before __yktrace_stop_tracing call.
+                    Tracing = false;
+                }
+            }
+            if (!Tracing && !llvm::isa<AllocaInst>(I)) {
+                // FIXME Remove allocas not needed in traced code.
```

Comment:
Probably done by LLVM itself already.

Reply:
Possibly. But it looks like we'll be generating all allocas ourselves anyway soon, so this will go away.